pkg/sql/distsql_running.go

```diff
@@ -1280,18 +1280,33 @@ func (dsp *DistSQLPlanner) planAndRunSubquery(
 			if row == nil {
 				break
 			}
+			var toAppend tree.Datum
 			if row.Len() == 1 {
 				// This seems hokey, but if we don't do this then the subquery expands
 				// to a tuple of tuples instead of a tuple of values and an expression
 				// like "k IN (SELECT foo FROM bar)" will fail because we're comparing
 				// a single value against a tuple.
-				result.D = append(result.D, row[0])
+				toAppend = row[0]
 			} else {
-				result.D = append(result.D, &tree.DTuple{D: row})
+				toAppend = &tree.DTuple{D: row}
 			}
+			// Perform memory accounting for this datum. We do this in an
+			// incremental fashion since we might be materializing a lot of data
+			// into a single result tuple, and the memory accounting below might
+			// come too late.
+			size := int64(toAppend.Size())
+			alreadyAccountedFor += size
+			if err = subqueryResultMemAcc.Grow(ctx, size); err != nil {
```

Comment:
Shouldn't the row container already have a total size of the data it's storing? If it does, then we can skip doing this extra work of re-sizing all datums as we pull them from the iterator, no? Like - we could simply account for that total size, again, up front, before we begin materializing into the tuple.